### ISSUE TITLE
Vite + Storybook 6.3 + Jest: "global is not defined"

### ISSUE THREAD

**Issue by adiun** [storybookjs/storybook#15391]
**Describe the bug**
I am trying to get Vite + Storybook 6.3 working with some existing stories. These stories use `jest.fn()` as mocks. I previously added this in `preview.js` to get `jest.fn()` working within storybook:
```
import jest from "jest-mock";
window.jest = jest;
```

But when using the Vite integration I am now getting an error when running storybook:
```
Uncaught ReferenceError: global is not defined
    at ../node_modules/jest-mock/build/index.js (index.js:948)
    at __require2 (chunk-IHTDASF6.js?v=3b66b68d:17)
    at dep:jest-mock:1
```

**To Reproduce**
I am trying `vite` with storybook 6.3 and npm 7.19 workspaces. My repro repo is here: https://github.com/adiun/vite-monorepo:
1. Ensure you have `npm 7.19` installed for the workspaces functionality
2. In the root, run `npm i`
3. In the `app` folder, run `npm run storybook`. 
4. In the browser console for storybook you will see:
```
Uncaught ReferenceError: global is not defined
    at ../node_modules/jest-mock/build/index.js (index.js:948)
    at __require2 (chunk-IHTDASF6.js?v=3b66b68d:17)
    at dep:jest-mock:1
```

**System**
```
Environment Info:

  System:
    OS: macOS 11.3
    CPU: (16) x64 Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
  Binaries:
    Node: 14.16.0 - /usr/local/bin/node
    npm: 7.19.0 - /usr/local/bin/npm
  Browsers:
    Chrome: 91.0.4472.114
    Edge: 91.0.864.48
    Firefox: 88.0.1
    Safari: 14.1
  npmPackages:
    @storybook/addon-docs: 6.3.0 => 6.3.0 
```

**Comment by adiun** [comment]
I was able to get a little further by mapping `global` to `window` in my `main.js`:
```
  async viteFinal(config) {
    return {
      ...config,
      define: {
        ...config.define,
        global: "window",
      },
    };
  }
```

Storybook loads now, but now I get this error:
`Cannot read property 'fn' of undefined` (the `jest` object which should be coming from `jest-mock` is `undefined` here)

![image](https://user-images.githubusercontent.com/120026/123559740-c8239f80-d752-11eb-8dbe-294678ff6f1a.png)

Note I also tried adding `jest-mock` to `optimizeDeps` and that didn't have an effect either.

**Comment by adiun** [comment]
Alright, I finally fixed this... `jest-mock` doesn't explicitly provide a default export so in storybook's `preview.js` this is what I have:

```javascript
import * as jest from "jest-mock";
window.jest = jest;
```

This combined with assigning `global` to `window` in `main.js` fixed this issue. 

**preview.js**
```javascript
import * as jest from "jest-mock";
window.jest = jest;

export const parameters = {
  actions: { argTypesRegex: "^on[A-Z].*" },
  controls: {
    matchers: {
      color: /(background|color)$/i,
      date: /Date$/,
    },
  },
};
```

**main.js**
```javascript
module.exports = {
  stories: ["../src/**/*.stories.mdx", "../src/**/*.stories.@(js|jsx|ts|tsx)"],
  addons: ["@storybook/addon-links", "@storybook/addon-essentials"],
  core: {
    builder: "storybook-builder-vite",
  },
  async viteFinal(config) {
    return {
      ...config,
      define: {
        ...config.define,
        global: "window",
      },
      esbuild: {
        ...config.esbuild,
        jsxInject: `import React from 'react'`,
      },
    };
  },
};

```

Updated [my repo](https://github.com/adiun/vite-monorepo) too

**Comment by shilman** [comment]
cc @eirslett

**Comment by adiun** [comment]
Following up on this... it's actually a bad idea to redefine global like this in `main.js`:

```js
define: {
  ...config.define,
  global: "window",
},
```

The [Vite docs even mention this](https://vitejs.dev/config/#define).

The problem is that when running a `build-storybook`, any references to `global` in libraries like `@storybook/client-logger` will be rewritten to `window` and it will break since it's running in `node`. 

So I removed this code in `main.js` to get `build-storybook` working and added this hack to `preview.js` to get `start-storybook` working:

```js
// Hack for jest-mock
window.global = window;
```

**main.js**:
```js
module.exports = {
  stories: ["../src/**/*.stories.mdx", "../src/**/*.stories.@(js|jsx|ts|tsx)"],
  addons: ["@storybook/addon-links", "@storybook/addon-essentials"],
  core: {
    builder: "storybook-builder-vite",
  },
  async viteFinal(config) {
    return {
      ...config,
      esbuild: {
        ...config.esbuild,
        jsxInject: `import React from 'react'`,
      },
      rollupOptions: {
        ...config.rollupOptions,
        // Externalize deps that shouldn't be bundled
        external: ["react", "react-dom"],
        output: {
          // Global vars to use in UMD build for externalized deps
          globals: {
            react: "React",
            "react-dom": "ReactDOM",
          },
        },
      },
    };
  },
};
```

**preview.js**:
```js
import * as jest from "jest-mock";
window.jest = jest;

// Hack for jest-mock
window.global = window;

export const parameters = {
  actions: { argTypesRegex: "^on[A-Z].*" },
  controls: {
    matchers: {
      color: /(background|color)$/i,
      date: /Date$/,
    },
  },
};
```

**Comment by shilman** [comment]
cc @eirslett

**Comment by eirslett** [comment]
Maybe we should document this in the README - a section about troubleshooting and potential workarounds? We could inject `window.global = window` into the iframe, generated by the builder itself, but I suspect that we may encounter edge cases where that's not desirable?
I guess the problem here is that Jest is written to be run in Node.js, where `global` is a variable. I suppose the "correct" solution would be to fix Jest, so it uses [globalThis](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/globalThis)?

**Comment by IanVS** [comment]
FWIW, jest [does now use globalThis](https://github.com/facebook/jest/pull/12447), but only in the alpha 28 versions so far.

**Comment by tobiasdiez** [comment]
Jest v28 is now released.

I've encountered a similar issue in the context of nuxt3, where the `global` imports in https://github.com/storybookjs/storybook/blob/next/app/vue3/src/client/preview/globals.ts make problems to hoist the preview, since depending on its configuration nuxt will render the page either on the serverside (node) or in the browser.

To support these use cases, it would be nice if globalThis is used everywhere. Can this issue maybe be reopened?

**Comment by IanVS** [comment]
@tobiasdiez I'm not familiar with nuxt, but as I understand, using the `global` npm package should work because it tests whether it is running in an environment with `global` or `window`.  https://github.com/Raynos/global/blob/master/window.js

Do you have a minimal reproduction we could take a look at?  Is your issue related to jest, or the import you linked in vue3?

**Comment by tobiasdiez** [comment]
Nuxt tries to cover a few distribution and render channels (e.g. service workers) and for this reason adds its own polyfills, but `global` in a cjs file is not yet supported https://github.com/nuxt/framework/issues/1922. One can workaround this issue by an adding custom alias that essentially redirects it to `globalThis`, see the instructions at https://github.com/nuxt-community/storybook/pull/385. But it would be nice if this workaround wouldn't be needed at all and storybook would rely directly on the more modern `globalThis`.

**Comment by IanVS** [comment]
I think maybe it is worth opening a new issue proposing a move to globalThis.

**Comment by Juice10** [comment]
For anyone using yarn, overriding the `jest-mock` version using `resolutions` in your package.json might work for you to get the latest version of `jest-mock` which doesn't have the `global` issue.
```json
"resolutions": {
    "jest-mock": "^28.1.0"
  },
```

**Comment by larsenwork** [comment]
> For anyone using yarn, overriding the `jest-mock` version using `resolutions`

Or with [overrides](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#overrides) in npm

**Comment by csantos1113** [comment]
> For anyone using yarn, overriding the `jest-mock` version using `resolutions` in your package.json might work for you to get the latest version of `jest-mock` which doesn't have the `global` issue.
> 
> ```json
> "resolutions": {
>     "jest-mock": "^28.1.0"
>   },
> ```

I'm confused, if this is the fix, couldn't we open a PR here in storybook to upgrade the dependency

```
├─ @storybook/addon-interactions@npm:7.0.0-alpha.41
│  └─ jest-mock@npm:27.4.0 (via npm:^27.0.6)
```

Or what else are we missing?

**Comment by IanVS** [comment]
Yes, I've been working on upgrading jest in the storybook packages, but it's a bit complex because of the way that jest types are global and the interactions with jest-dom.  I'm waiting for some feedback on https://github.com/testing-library/jest-dom/pull/483.

**Comment by lamualfa** [comment]
Overriding `jest-mock` version doesn't work for me even after deleting `node_modules` folder.

[Override](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#overrides) field in `package.json` file:
```json
{
 "overrides": {
    "@storybook/addon-interactions": {
      "jest-mock": "^29.3.1"
    }
  }
}
```

Output of `npm list jest-mock`:
```bash
└─┬ @storybook/addon-interactions@6.5.13
  └── jest-mock@29.3.1 overridden
```

Any idea about this problem @Juice10 @larsenwork?

**Comment by bryanjtc** [comment]
Any update on this?

**Comment by valentinpalkovic** [comment]
Thanks for the report! v9 added a first-party mocking API (`fn()`, `spyOn()` from `storybook/test`), removing the need to pull in `jest-mock`. The vite builder has also been rewritten with proper `global` handling. Closing as resolved by the v9 mocking API.
